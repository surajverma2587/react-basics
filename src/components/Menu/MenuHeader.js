import React from 'react';
import PropTypes from 'prop-types';
/* external library to render the stars for ratings */
/* https://www.npmjs.com/package/react-star-ratings */
import StarRatings from 'react-star-ratings';

const MenuHeader = (/* props object */) => {
  /* bind the data from the props object to the html template */
  /* replace the bottom return statement with the populated html template */
  return (
    <div>
      Menu Header Component
    </div>
  );
};

/* display name for the component */
MenuHeader.displayName = 'MenuHeader';

MenuHeader.propTypes = {
  /* add Proptypes here */
};

export default MenuHeader;