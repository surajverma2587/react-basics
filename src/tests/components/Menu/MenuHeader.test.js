import React from 'react';
import { shallow } from 'enzyme';

import MenuHeader from '../../../components/Menu/MenuHeader';

const props = {
  title: 'Menu title',
  address: 'Menu address',
  imgUrl: 'https://google.com',
  reviews: 100,
  rating: 2.5,
};

describe('Menu Header Component', () => {
  describe('when props are valid', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<MenuHeader {...props} />);
    });
    
    it('should render snapshot of the component', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should render the component', () => {
      const menuHeader = wrapper.find('.media');
      expect(menuHeader.exists()).toBe(true);
    });

    it('should render the image in the menu header', () => {
      const img = wrapper.find('img');
      expect(img.prop('src')).toEqual(props.imgUrl);
      expect(img.prop('alt')).toEqual(props.title);
    });
  
    it('should render the title of the restaurant', () => {
      const title = wrapper.find('h2').at(0);
      expect(title.text()).toEqual(props.title);
    });
  
    it('should render the reviews of the restaurant', () => {
      const reviews = wrapper.find('span').at(0);
      expect(reviews.text()).toEqual(`${props.reviews} reviews`);
    });

    it('should render the address of the restaurant', () => {
      const address = wrapper.find('span').at(1);
      expect(address.text()).toEqual(props.address);
    });
  
    it('should render the StarRatings component', () => {
      const starRatings = wrapper.find('StarRatings');
      expect(starRatings.prop('rating')).toEqual(props.rating);
    });
  });

  describe('when props are invalid', () => {
    it('should not render the component when props are not present', () => {
      const wrapper = shallow(<MenuHeader />);  
      expect(wrapper.find('.media').exists()).toBe(false);
    });

    it('should not render the component when any required prop is invalid', () => {
      const invalidProps = [
        { ...props, title: null },
        { ...props, address: null },
      ];
      invalidProps.forEach(invalidProp => {
        const wrapper = shallow(<MenuHeader {...invalidProp} />);
        expect(wrapper.find('.media').exists()).toBe(false);
      });
    });

    it('should render 0 reviews of the restaurant when reviews are not available', () => {
      const noReviewProps = {
        id: 'A',
        title: 'Menu title',
        address: 'Menu address',
        imgUrl: 'https://google.com',
        rating: 2.5,
        menuSections: props.menuSections,
        onClickMenuItem: jest.fn(),
      }
      const wrapper = shallow(<MenuHeader {...noReviewProps} />);
      const reviews = wrapper.find('span').at(0);
      expect(reviews.text()).toEqual('0 reviews');
    });

    it('should render the StarRatings component with 0 stars when ratings is not available', () => {
      const noRatingsProps = {
        id: 'A',
        title: 'Menu title',
        address: 'Menu address',
        imgUrl: 'https://google.com',
        reviews: 22,
        menuSections: props.menuSections,
        onClickMenuItem: jest.fn(),
      };
      const wrapper = shallow(<MenuHeader {...noRatingsProps} />);
      const starRatings = wrapper.find('StarRatings');
      expect(starRatings.prop('rating')).toEqual(0);
    });
  });
});
