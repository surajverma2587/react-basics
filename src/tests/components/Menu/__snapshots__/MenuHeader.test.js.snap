// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Header Component when props are valid should render snapshot of the component 1`] = `
<div
  className="media"
>
  <div
    className="media-left"
  >
    <img
      alt="Menu title"
      className="mr-3"
      height={100}
      src="https://google.com"
      width={120}
    />
  </div>
  <div
    className="media-body"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-12"
      >
        <h2>
          Menu title
        </h2>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-2"
      >
        <StarRatings
          changeRating={null}
          gradientPathName=""
          ignoreInlineStyles={false}
          numberOfStars={5}
          rating={2.5}
          starDimension="20px"
          starEmptyColor="rgb(203, 211, 227)"
          starHoverColor="rgb(230, 67, 47)"
          starRatedColor="red"
          starSpacing="1px"
          svgIconPath="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
          svgIconViewBox="0 0 51 48"
          typeOfWidget="Star"
        />
      </div>
      <div
        className="col-sm-10"
      >
        <span>
          100
           reviews
        </span>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-12"
      >
        <span>
          Menu address
        </span>
      </div>
    </div>
  </div>
</div>
`;
